pkg/kubelet/cm/dra/manager.go

```diff
@@ -302,34 +335,58 @@ func (m *ManagerImpl) UnprepareResources(pod *v1.Pod) error {
 				pluginName = claimInfo.DriverName
 			}
 
-			// Call NodeUnprepareResource RPC for each resourceHandle
-			client, err := dra.NewDRAPluginClient(pluginName)
-			if err != nil {
-				return fmt.Errorf("failed to get DRA Plugin client for plugin name %s, err=%+v", pluginName, err)
+			batches[pluginName] = append(batches[pluginName],
+				&drapb.Claim{
+					Namespace:      resourceClaim.Namespace,
+					Uid:            string(resourceClaim.UID),
+					Name:           resourceClaim.Name,
+					ResourceHandle: resourceHandle.Data,
+				})
+		}
+		claimInfos[resourceClaim.UID] = claimInfo
+	}
+
+	// Check for errors. If there is any error, processing gets aborted.
+	// We could try to continue, but that would make the code more complex.
+	for pluginName, claims := range batches {
+		// Call NodeUnprepareResources RPC for all resource handles.
+		client, err := dra.NewDRAPluginClient(pluginName)
+		if err != nil {
+			return fmt.Errorf("failed to get DRA Plugin client for plugin name %s: %v", pluginName, err)
+		}
+		response, err := client.NodeUnprepareResources(context.Background(), &drapb.NodeUnprepareResourcesRequest{Claims: claims})
+		if err != nil {
+			// General error unrelated to any particular claim.
+			return fmt.Errorf("NodeUnprepareResources failed: %v", err)
+		}
+
+		for claimUID, result := range response.Claims {
+			reqClaim := lookupClaimRequest(claims, claimUID)
+			if reqClaim == nil {
+				return fmt.Errorf("NodeUnprepareResources returned result for unknown claim UID %s", claimUID)
 			}
-			response, err := client.NodeUnprepareResource(
-				context.Background(),
-				claimInfo.Namespace,
-				claimInfo.ClaimUID,
-				claimInfo.ClaimName,
-				resourceHandle.Data)
-			if err != nil {
-				return fmt.Errorf(
-					"NodeUnprepareResource failed, pod: %s, claim UID: %s, claim name: %s, resource handle: %s, err: %+v",
-					pod.Name, claimInfo.ClaimUID, claimInfo.ClaimName, resourceHandle.Data, err)
+			if result.Error != "" {
+				return fmt.Errorf("NodeUnprepareResources failed for claim %s/%s: %s", reqClaim.Namespace, reqClaim.Name, err)
 			}
-			klog.V(3).InfoS("NodeUnprepareResource succeeded", "response", response)
 		}
 
-		// Delete last pod UID only if all NodeUnprepareResource calls succeed.
+		unfinished := len(claims) - len(response.Claims)
+		if unfinished != 0 {
+			return fmt.Errorf("NodeUnprepareResources left out %d claims", unfinished)
+		}
+	}
+
+	// No errors, time to remove the pod reference from all claims that were involved.
+	for _, claimInfo := range claimInfos {
+		// Delete last pod UID only if unprepare succeed.
```

Comment:
This is a slight change in semantics from before, since previously we would remove pod references as `NodeUnprepareResources()` was called successfully on each individual claim. I'm not saying it's not OK to do it this new way, but we should acknowledge that it's a slight change in semantics from before.

Reply:
Let's keep the original semantic. I moved this code into the for loop above, directly after checking the per-claim result.